File: src/main/java/com/github/pnemonic/game/lottery/lotto/LottoTester.java
package com.github.pnemonic.game.lottery.lotto;

import com.github.pnemonic.game.NumberStatistic;
import com.github.pnemonic.game.NumberStatisticGrouping;
import com.github.pnemonic.game.Tester;
import com.github.pnemonic.game.lottery.LotteryGame;
import com.github.pnemonic.game.lottery.LotteryRecord;
import com.github.pnemonic.game.lottery.LotteryResultsReader;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Test various strategies for "Lotto".
 *
 * @author moshew
 */
public class LottoTester extends Tester {

    // per hundred
    private static final int THRESHOLD_CANDIDATES_PERCENT = 50;

    private static final int WIN = 601;
    /**
     * Maximum repeat of same number.
     */
    private static final int MAX_REPEAT_THRESHOLD = 8;

    private final int thresholdCandidates;
    private final int thresholdCandidatesAnd;
    private final int thresholdCandidatesOr;
    private final int thresholdCandidatesOr3;
    private final int thresholdCandidatesOr4;
    private final List<Integer> candidates;

    /**
     * Creates a new tester.
     */
    public LottoTester() {
        super(new Lotto());
        thresholdCandidates = (numBalls * THRESHOLD_CANDIDATES_PERCENT) / 100;
        thresholdCandidatesAnd = Math.min((thresholdCandidates * 3) / 2, numBalls / 2);
        thresholdCandidatesOr = thresholdCandidates / 2;
        thresholdCandidatesOr3 = thresholdCandidates / 3;
        thresholdCandidatesOr4 = thresholdCandidates / 4;
        this.candidates = new ArrayList<Integer>(numBalls);

        if (thresholdCandidates < lotterySize) {
            throw new IllegalArgumentException("too few candidates");
        }
    }

    /**
     * Main method.
     *
     * @param args the array of arguments.
     */
    public static void main(String[] args) {
        String fileName = (args.length == 0) ? "results/Lotto.csv" : args[0];
        File file = new File(fileName);
        LottoTester tester = new LottoTester();
        try {
            tester.parse(file);
            tester.drive();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void parse(File file) throws IOException {
        LotteryResultsReader reader = new LottoResultsReader();
        this.records = reader.parse(file);
        this.recordsSize = records.size();
        this.numGamesTotal = recordsSize * Lotto.PLAYS;
    }

    @Override
    public void drive() {
        LottoStats stats = new LottoStats(new Lotto());
        try {
            stats.processRecords(records);
            this.numStats = stats.getNumberStatistics();

            for (NumberStatisticGrouping grouping : NumberStatisticGrouping.values()) {
                drive(grouping, grouping.name());
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void drive(NumberStatisticGrouping grouping, String name) {
        Set<LotteryGame> games;
        int score;
        int maxScore = 0;
        int totScore = 0;
        int win = 0;
        int recordIndex = 0;

        candidates.clear();

        for (LotteryRecord record : records) {
            lottery.setCandidates(candidates);

            games = play(Lotto.PLAYS);
            score = 0;
            for (LotteryGame game : games) {
                score = record.compareTo(game);
                totScore += score;
                maxScore = Math.max(score, maxScore);
            }
            win += (maxScore == WIN) ? 1 : 0;

            nextCandidates(grouping, recordIndex);

            recordIndex++;
        }
        int aveScore = totScore / numGamesTotal;
        float winPercent = ((float) win / recordsSize) * 100;
        System.out.println(name + ":\t{max. " + maxScore + ";\tave. " + aveScore + ";\twin " + winPercent + "%}");
    }

    /**
     * Use statistics to determine the next candidates.
     */
    protected void nextCandidates(NumberStatisticGrouping grouping, int row) {
        // Get the statistics.
        NumberStatistic[] nstatRow = numStats[row];
        candidates.clear();
        boolean add;

        for (NumberStatistic nstat : nstatRow) {
            add = nstat.repeat < MAX_REPEAT_THRESHOLD;
            switch (grouping) {
                case REGULAR:
                    break;
                case LEAST_REPEAT:
                    add &= (nstat.indexLeastRepeat < thresholdCandidates);
                    break;
                case MOST_REPEAT:
                    add &= (nstat.indexMostRepeat < thresholdCandidates);
                    break;
                case LC:
                    add &= (nstat.indexLeastCount < thresholdCandidates);
                    break;
                case LU:
                    add &= (nstat.indexLeastUsed < thresholdCandidates);
                    break;
                case LU_LC:
                    add &= (nstat.indexLeastUsed < thresholdCandidatesOr) || (nstat.indexLeastCount < thresholdCandidatesOr);
                    break;
                case LU_MC:
                    add &= (nstat.indexLeastUsed < thresholdCandidatesOr) || (nstat.indexMostCount < thresholdCandidatesOr);
                    break;
                case LU_MC_LC:
                    add &= (nstat.indexLeastUsed < thresholdCandidatesOr3) || (nstat.indexMostCount < thresholdCandidatesOr3)
                        || (nstat.indexLeastCount < thresholdCandidatesOr3);
                    break;
                case MC:
                    add &= (nstat.indexMostCount < thresholdCandidates);
                    break;
                case MC_LC:
                    add &= (nstat.indexMostCount < thresholdCandidatesOr) || (nstat.indexLeastCount < thresholdCandidatesOr);
                    break;
                case MU:
                    add &= (nstat.indexMostUsed < thresholdCandidates);
                    break;
                case MU_LC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr) || (nstat.indexLeastCount < thresholdCandidatesOr);
                    break;
                case MU_LU:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr) || (nstat.indexLeastUsed < thresholdCandidatesOr);
                    break;
                case MU_LU_LC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr3) || (nstat.indexLeastUsed < thresholdCandidatesOr3)
                        || (nstat.indexLeastCount < thresholdCandidatesOr3);
                    break;
                case MU_LU_MC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr3) || (nstat.indexLeastUsed < thresholdCandidatesOr3)
                        || (nstat.indexMostCount < thresholdCandidatesOr3);
                    break;
                case MU_LU_MC_LC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr4) || (nstat.indexLeastUsed < thresholdCandidatesOr4)
                        || (nstat.indexMostCount < thresholdCandidatesOr4) || (nstat.indexLeastCount < thresholdCandidatesOr4);
                    break;
                case MU_MC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr) || (nstat.indexMostCount < thresholdCandidatesOr);
                    break;
                case MU_MC_LC:
                    add &= (nstat.indexMostUsed < thresholdCandidatesOr3) || (nstat.indexMostCount < thresholdCandidatesOr3)
                        || (nstat.indexLeastCount < thresholdCandidatesOr3);
                    break;
                case LC_AND_LU:
                    add &= (nstat.indexLeastCount < thresholdCandidatesAnd) && (nstat.indexLeastUsed < thresholdCandidatesAnd);
                    break;
                case LC_AND_MU:
                    add &= (nstat.indexLeastCount < thresholdCandidatesAnd) && (nstat.indexMostUsed < thresholdCandidatesAnd);
                    break;
                case MC_AND_LU:
                    add &= (nstat.indexMostCount < thresholdCandidatesAnd) && (nstat.indexLeastUsed < thresholdCandidatesAnd);
                    break;
                case MC_AND_MU:
                    add &= (nstat.indexMostCount < thresholdCandidatesAnd) && (nstat.indexMostUsed < thresholdCandidatesAnd);
                    break;
                default:
                    System.err.println("unhandled type " + grouping);
                    break;
            }
            if (add) {
                candidates.add(nstat.id);
            }
        }
    }
}

File: src/main/java/com/github/pnemonic/game/lottery/lotto/LottoTester.kt
package com.github.pnemonic.game.lottery.lotto

import com.github.pnemonic.game.NumberStatisticGrouping
import com.github.pnemonic.game.lottery.LotteryGame
import com.github.pnemonic.game.lottery.LotteryResultsReader
import com.github.pnemonic.game.lottery.LotteryTester
import com.github.pnemonic.game.lottery.pais777.Lotto777.Companion.COST
import com.github.pnemonic.game.lottery.pais777.Lotto777Tester.Companion.BUDGET
import java.io.File
import kotlin.math.max
import kotlin.math.min

/**
 * Test various strategies for "Lotto".
 */
class LottoTester : LotteryTester<Lotto, LottoStats>(Lotto()) {
    private val thresholdCandidates: Int = (numBalls * THRESHOLD_CANDIDATES_PERCENT) / 100
    private val thresholdCandidatesOr: Int = thresholdCandidates / 2

    init {
        require(thresholdCandidates >= lotterySize) { "too few candidates" }
    }

    override fun createResultsReader(): LotteryResultsReader {
        return LottoResultsReader()
    }

    override fun createStats(lottery: Lotto): LottoStats {
        return LottoStats(lottery)
    }

    override fun drive(grouping: NumberStatisticGrouping, stats: LottoStats) {
        // Total number of plays per budget. Lotto played in pairs.
        val numPlays = (BUDGET / COST) and 1.inv()
        var games: Collection<LotteryGame>
        var score: Int
        var maxScore = 0
        var totalScore = 0
        var win = 0
        var recordIndex = -1
        var numGamesTotal = 0
        for (record in records) {
            predictNextCandidates(grouping, stats, recordIndex)
            games = play(numPlays, record)
            for (game in games) {
                score = record.compareTo(game)
                totalScore += score
                maxScore = max(score, maxScore)
            }
            win += if (maxScore == WIN) 1 else 0
            numGamesTotal += games.size
            recordIndex++
        }
        val aveScore = totalScore / numGamesTotal
        val winPercent = (win * 100f) / records.size
        println("$grouping:\t{max. $maxScore;\tave. $aveScore;\twin $winPercent%}")
    }

    /**
     * Use statistics to determine the next candidates.
     */
    private fun predictNextCandidates(grouping: NumberStatisticGrouping, stats: LottoStats, row: Int) {
        if (row < 0) {
            lottery.setCandidates(null)
            return
        }
        // Get the statistics.
        val numStats = stats.numberStatistics
        val nstatRow = numStats[row]
        val candidates = mutableListOf<Int>()
        var add: Boolean
        for (nstat in nstatRow) {
            add = (nstat.repeat < MAX_REPEAT_THRESHOLD)
            when (grouping) {
                NumberStatisticGrouping.LEAST_COUNT ->
                    add = add && (nstat.indexLeastCount < thresholdCandidates)

                NumberStatisticGrouping.LEAST_REPEAT ->
                    add = add && (nstat.indexLeastRepeat < thresholdCandidates)

                NumberStatisticGrouping.LEAST_USED ->
                    add = add && (nstat.indexLeastUsed < thresholdCandidates)

                NumberStatisticGrouping.MOST_COUNT ->
                    add = add && (nstat.indexMostCount < thresholdCandidates)

                NumberStatisticGrouping.MOST_REPEAT ->
                    add = add && (nstat.indexMostRepeat < thresholdCandidates)

                NumberStatisticGrouping.MOST_USED ->
                    add = add && (nstat.indexMostUsed < thresholdCandidates)

                NumberStatisticGrouping.LC_AND_LU -> {
                    add = add && (nstat.indexLeastCount < thresholdCandidates)
                    add = add && (nstat.indexLeastUsed < thresholdCandidates)
                }

                NumberStatisticGrouping.LC_AND_MU -> {
                    add = add && (nstat.indexLeastCount < thresholdCandidates)
                    add = add && (nstat.indexMostUsed < thresholdCandidates)
                }

                NumberStatisticGrouping.MC_AND_LU -> {
                    add = add && (nstat.indexMostCount < thresholdCandidates)
                    add = add && (nstat.indexLeastUsed < thresholdCandidates)
                }

                NumberStatisticGrouping.MC_AND_MU -> {
                    add = add && (nstat.indexMostCount < thresholdCandidates)
                    add = add && (nstat.indexMostUsed < thresholdCandidates)
                }

                NumberStatisticGrouping.LU_LC ->
                    add = add && (nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.LU_MC ->
                    add = add && (nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr)

                NumberStatisticGrouping.LU_MC_LC ->
                    add = add && (nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MC_LC ->
                    add = add && (nstat.indexMostCount < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_LC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_LU ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexLeastUsed < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_LU_LC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_LU_MC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_LU_MC_LC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexLeastUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_MC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr)

                NumberStatisticGrouping.MU_MC_LC ->
                    add = add && (nstat.indexMostUsed < thresholdCandidatesOr ||
                            nstat.indexMostCount < thresholdCandidatesOr ||
                            nstat.indexLeastCount < thresholdCandidatesOr)

                else -> Unit
            }
            if (add) {
                candidates.add(nstat.id)
            }
        }
        lottery.setCandidates(candidates)
    }

    companion object {
        // per hundred
        private const val THRESHOLD_CANDIDATES_PERCENT = 50
        private const val WIN = 601

        /**
         * Maximum repeat of same number.
         */
        private const val MAX_REPEAT_THRESHOLD = 8
    }
}

/**
 * Main method.
 *
 * @param args the array of arguments.
 */
fun main(args: Array<String>) {
    val fileName = if (args.isEmpty()) "results/Lotto.csv" else args[0]
    val file = File(fileName)
    val tester = LottoTester()
    try {
        tester.parse(file)
        tester.drive()
    } catch (e: Exception) {
        e.printStackTrace()
    }
}